### PR TITLE
Feature/raft fine tuning

### PR DESCRIPTION
**Pull request by efenocchi**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
This PR introduces the ability to load a dataset from Activeloop by calling `load_deep_lake_dataset` instead of `load_dataset` (see `torchtune/torchtune/datasets/_utils.py`). Additionally, it enables fine-tuning of available models using the RAFT technique (see `torchtune/recipes/configs/llama3/8B_lora_single_device_deep_lake_raft.yaml`).

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add unit tests for any new functionality
- [X] update docstrings for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [X] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

I want to report that there is an error in the main repository when running the integration tests in the `tests/recipes/test_eleuther_eval.py` file. 
integration_test_error_1
integration_test_error_2

### CODE REVIEW

File: torchtune/datasets/_instruct_raft.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    InstructTemplate,
+    Message,
+    validate_messages,
+)
+from torchtune.datasets._utils import load_deep_lake_dataset
+
+from torchtune.modules.tokenizers import Tokenizer
+
+
+class InstructDatasetDeepLakeRAFT(Dataset):
```
Comment (ebsmothers) [inline torchtune/datasets/_instruct_raft.py:23]: High-level question on this class: is the only difference between this and `InstructDataset` the usage of `load_deep_lake_dataset` instead of `load_dataset`? If so, I wonder if this is something we should consider parametrizing rather than having to write an entirely new dataset class. cc @RdoubleA for any thoughts here

File: torchtune/datasets/_raft.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from torchtune.data import RAFTInstructTemplate
+from torchtune.datasets._instruct_raft import InstructDatasetDeepLakeRAFT
+from torchtune.modules.tokenizers import Tokenizer
+
+activeloop_dataset = "hub://manufe/raft_format_dataset_biomedical"  # Replace with your ActiveLoop dataset
+
+
+def raft_dataset(
```
Comment (ebsmothers) [inline torchtune/datasets/_raft.py:14]: What is this dataset exactly? Seems like `raft_dataset` is still the generic dataset type. But for dataset builders like these we typically point to a canonical dataset on the Hub (e.g. Alpaca). Is `raft_format_dataset_biomedical` the canonical dataset here? If so, is this something that's widely-used, or intended more as a demo?

**Review by ebsmothers — COMMENTED** [review]
Thanks for the PR! Appreciate you adding the unit tests as well. My main questions are around the addition of the new dependency and its handling on imports, need for a dataloader class, and the new configs. 

Re the failing test, we do not currently have lm_eval in our optional dependencies due to some other transitive dependencies, so if you follow the suggestion in the console about `pip install "lm_eval=0.4.*"` I suspect you will be able to get that test to pass.

File: pyproject.toml [hunk 3]
```diff
@@ -47,6 +47,7 @@ dev = [
     "pytest-integration",
     "tensorboard",
     "wandb",
+    "deeplake"
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply pyproject.toml:50]: if I'm not mistaken deeplake requires : aioboto3, boto3, click, humbug, libdeeplake, lz4, nest-asyncio, numpy, pathos, pillow, pydantic, pyjwt, tqdm (tested with pip show deeplake, tell me if something is wrong)

File: recipes/configs/llama2/7B_lora_single_device_deep_lake_raft.yaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# Config for single device LoRA finetuning in lora_finetune_single_device.py
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply recipes/configs/llama2/7B_lora_single_device_deep_lake_raft.yaml:1]: I thought it would be helpful to have an example because it's not just another data set but it's another technique to fine tune with. I thought this example would make it easier to use this technique

File: torchtune/data/_instruct_templates.py [hunk 5]
```diff
@@ -87,6 +87,43 @@ def format(
         return prompt
 
 
+class RAFTInstructTemplate(InstructTemplate):
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply torchtune/data/_instruct_templates.py:90]: The difference is that RAFT by default has different column names and in this way we can manage it without having to make changes at the dataset level

File: torchtune/datasets/_instruct_raft.py (hunk 1 above)
Comment (efenocchi) [reply torchtune/datasets/_instruct_raft.py:23]: yes I think it can be parameterized, can you give me some advice on how we can proceed here?

File: torchtune/datasets/_raft.py (hunk 2 above)
Comment (efenocchi) [reply torchtune/datasets/_raft.py:14]: RAFT being a fairly new technique there is no real reference dataset, this is a dataset that we are using on a real project and can be used as a demo

File: torchtune/datasets/_utils.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import deeplake
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply torchtune/datasets/_utils.py:7]: How can we handle this part?

File: torchtune/datasets/_utils.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+#         ds (deeplake.Dataset): The dataset object obtained from ActiveLoop's DeepLake platform.
+
+# Example:
+#     # Load a dataset from DeepLake and create a DataLoader
+#     dataset = load_deeplake_dataset("my_deep_lake_dataset")
+#     dataloader = DeepLakeDataloader(dataset)
+# """
+
+
+class DeepLakeDataloader(Dataset):
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply torchtune/datasets/_utils.py:36]: To get the dataset values ​​we need to create a custom dataset otherwise we would not be able to access the data without throwing any error

File: recipes/lora_finetune_single_device.py [hunk 8]
```diff
@@ -525,4 +525,4 @@ def recipe_main(cfg: DictConfig) -> None:
 
 
 if __name__ == "__main__":
-    sys.exit(recipe_main())
+    sys.exit(recipe_main())
```
[the remark this answers is not in the document]
Comment (efenocchi) [reply recipes/lora_finetune_single_device.py]: Checked again, added a white line at end of file, after all tests passed again

**Comment by efenocchi** [comment]
Hi @ebsmothers, 
thank you very much for the suggestions and for the quick response you gave me, sorry it took me so long to reply.. if there is anything I can do to get this PR accepted I will be happy to do it.
Have a good weekend.
Emanuele

File: torchtune/datasets/_raft.py (hunk 2 above)
Comment (efenocchi) [reply torchtune/datasets/_raft.py:14]: here too perhaps we can replace the dataset with an input taken in the yaml file or passed as a parameter in the cli

File: torchtune/data/_instruct_templates.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (efenocchi) [reply torchtune/data/_instruct_templates.py:90]: I just refined the prompt to closely align with the structure outlined in the referenced paper

**Review by ebsmothers — COMMENTED** [review]
Hi @efenocchi thanks for making the updates and apologies for the delay on getting back to you. Given that 

(a) we try to keep our core dependencies pretty minimal in torchtune, and 
(b) integrating a dev dependency tightly into our core code (in this case `torchtune/datasets`) is something we try to avoid,

I think we may want to take a slightly different approach on this PR. We have another ongoing integration with ClearML Logger and I wonder if we can follow a similar path here. Namely, you can push the change to a fork, and we can highlight it as a community integration in our README. Then as we get more usage and requests to integrate we can look at merging into main. Basically the process outlined [here](https://github.com/pytorch/torchtune/pull/895#issuecomment-2088809146).

Let me know if this type of integration makes sense to you. Thanks!

**Comment by efenocchi** [comment]
Hi @ebsmothers, 
Thank you for your feedback and for outlining the approach.
I understand the importance of keeping core dependencies minimal in torchtune and avoiding tight integration of development dependencies within the core code. I think it's a good plan to follow the same approach as the ClearML Logger integration.

I've updated the code in my fork, let me know if there's anything else you'd like me to adjust or if there are further steps you'd like me to take. 

Thanks again!
